src/java/org/apache/cassandra/metrics/ThreadLocalMetrics.java

```diff
@@ -0,0 +1,352 @@
+            phantomReferences.remove(obj);
+        }
+    }
+
+    private static class MetricIdReference extends PhantomReference<Object>
+    {
+        private final int metricId;
+
+        public MetricIdReference(Object referent, ReferenceQueue<? super Object> q, int metricId)
+        {
+            super(referent, q);
+            this.metricId = metricId;
+        }
+
+        public void release()
+        {
+            recycleMetricId(metricId);
+        }
+    }
+
+    private static class MetricCleanerReference extends PhantomReference<Object>
+    {
+        private final MetricCleaner metricCleaner;
+
+        public MetricCleanerReference(Object referent, ReferenceQueue<? super Object> q, MetricCleaner metricCleaner)
+        {
+            super(referent, q);
+            this.metricCleaner = metricCleaner;
+        }
+
+        public void release()
+        {
+            metricCleaner.clean();
+        }
+    }
+
+    interface MetricCleaner
+    {
+        void clean();
+    }
+
+    static void destroyWhenUnreachable(Object referent, int metricId)
+    {
+        phantomReferences.add(new MetricIdReference(referent, referenceQueue, metricId));
+    }
+
+    static void destroyWhenUnreachable(Object referent, MetricCleaner metricCleaner)
+    {
+        phantomReferences.add(new MetricCleanerReference(referent, referenceQueue, metricCleaner));
+    }
+
+    @VisibleForTesting
+    public static void shutdownCleaner(long timeout, TimeUnit unit) throws InterruptedException, TimeoutException
+    {
+        shutdownAndWait(timeout, unit, of(cleaner));
+    }
+
+    private void release()
+    {
+        Lock lock = summaryLock.writeLock();
+        lock.lock();
+        try
+        {
+            // we may try to release ThreadLocalMetrics 2 times: onRemoval and by PhantomReference
+            // so this if check is needed to avoid a potential double release
+            if (allThreadLocalMetrics.remove(this))
+                for (int metricId = 0; metricId < counterValues.length; metricId++)
+                {
+                    long value = counterValues[metricId];
+                    if (value != 0)
+                        getSummary(metricId).addAndGet(value);
+                }
+        }
+        finally
+        {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * If we already have ThreadLocalMetrics instance looked up for the current thread
+     * we can use this method to avoid thread local lookup costs.
+     * It can be used if you need to update several counters at the same time.
+     * @param metricId metric to add a value
+     * @param n valuen to add, can be negative number as well
+     */
+    public void addNonStatic(int metricId, long n)
+    {
+        getNonStatic(metricId)[metricId] += n;
+    }
+
+    public static void add(int metricId, long n)
+    {
+        get(metricId)[metricId] += n;
+    }
+
+    private static long getCount(int metricId, boolean resetToZero)
```

Comment:
Should we consider a batch version of this where we compute a snapshot of the summary+thread local for all metrics, and refresh this if older than some number of seconds and otherwise serve from the snapshot?